bchrpc/server.go

```diff
@@ -1113,16 +1208,421 @@ func (s *GrpcServer) GetMerkleProof(ctx context.Context, req *pb.GetMerkleProofR
+	}
+
+	extK, err := ext.Child(req.AddressIndex)
+	if err != nil {
+		fmt.Println(err)
+		return nil, status.Error(codes.InvalidArgument, "invalid xpub: "+err.Error())
+	}
+
+	pubKey, _ := extK.ECPubKey()
+	addr, _ := extK.Address(s.chainParams)
+	slpAddrStr := ""
+	if s.slpIndex != nil {
+		slpAddr, _ := goslp.NewAddressPubKeyHash(addr.Hash160()[:], s.chainParams)
+		slpAddrStr = s.chainParams.SlpAddressPrefix + ":" + slpAddr.EncodeAddress()
+	}
+
+	res := &pb.GetBip44HdAddressResponse{
+		PubKey:   pubKey.SerializeCompressed(),
+		CashAddr: s.chainParams.CashAddressPrefix + ":" + addr.EncodeAddress(),
+		SlpAddr:  slpAddrStr,
+	}
+
+	return res, nil
+}
+
+func isMaybeSlpTransaction(txn *wire.MsgTx) bool {
+	if len(txn.TxOut) > 0 {
+		lokadHex, _ := hex.DecodeString("534c5000")
+		return bytes.Contains(txn.TxOut[0].PkScript, lokadHex)
+	}
+	return false
+}
+
+// CheckSlpTransaction checks validity of a submitted transaction and with return an error if the transaction is invalid
+func (s *GrpcServer) CheckSlpTransaction(ctx context.Context, req *pb.CheckSlpTransactionRequest) (*pb.CheckSlpTransactionResponse, error) {
+
+	if req.RequiredSlpBurns != nil {
+		return nil, status.Error(codes.Aborted, "intentional burning with 'AllowedSlpBurns' is not yet implemented")
+	}
+
+	msgTx := &wire.MsgTx{}
+	if err := msgTx.BchDecode(bytes.NewReader(req.Transaction), wire.ProtocolVersion, wire.BaseEncoding); err != nil {
+		return nil, status.Error(codes.InvalidArgument, "unable to deserialize transaction")
+	}
+
+	if len(msgTx.TxIn) == 0 || len(msgTx.TxOut) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "transaction is missing inputs or outputs")
+	}
+
+	isMaybeSlp := isMaybeSlpTransaction(msgTx)
+	if !isMaybeSlp {
+		return nil, status.Error(codes.Aborted, "invalid slp (lokad id is missing)")
+	}
+
+	err := s.checkSlpTransaction(msgTx)
+	if err != nil {
+		return nil, err
+	}
+
+	res := &pb.CheckSlpTransactionResponse{
+		IsValid: true,
+	}
+	return res, nil
+}
+
+func (s *GrpcServer) checkSlpTransaction(msgTx *wire.MsgTx) error {
+	if s.slpIndex == nil {
+		return status.Error(codes.Unavailable, "slpindex required")
+	}
+
+	slpMsg, err := v1parser.ParseSLP(msgTx.TxOut[0].PkScript)
+	if err != nil {
+		return status.Error(codes.Aborted, "submitted transaction rejected to prevent token burn (error parsing slp op_return message: "+err.Error()+")")
+	}
+
+	if slpMsg.TransactionType == "SEND" {
+		// get token ID from slpindex
+		tokenIDHash := slpMsg.Data.(v1parser.SlpSend).TokenID
+
+		// loop through **ALL** inputs, accumulate input amount for tokenID, abort on SLP input with wrong ID
+		inputVal := big.NewInt(0)
+		for _, txIn := range msgTx.TxIn {
+			var entry *indexers.SlpIndexEntry
+
+			entry, err = s.getSlpIndexEntry(&txIn.PreviousOutPoint.Hash)
+			if err != nil {
+				continue
+			}
+			if !bytes.Equal(tokenIDHash, entry.TokenIDHash.CloneBytes()) {
+				// TODO: (BONUS) check 'req.RequiredSlpBurns' for allowed burns
```

Comment:
Lets try to get this implemented. =)

Reply:
will do, I have it partially implemented in my stash :)